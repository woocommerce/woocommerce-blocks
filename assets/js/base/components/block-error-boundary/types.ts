interface BlockErrorBase {
	/**
	 * URL of the image to display.
	 * If it's `null` or an empty string, no image will be displayed.
	 * If it's not defined, the default image will be used.
	 */
	imageUrl?: string;
	/**
	 * Text to display as the heading of the error block.
	 * If it's `null` or an empty string, no header will be displayed.
	 * If it's not defined, the default header will be used.
	 */
	header?: string;
	/**
	 * Text to display in the error block below the header.
	 * If it's `null` or an empty string, nothing will be displayed.
	 * If it's not defined, the default text will be used.
	 */
	text: React.ReactNode;
	/**
	 * Text preceeding the error message.
	 */
	errorMessagePrefix?: string;
	/**
	 * Button cta.
	 */
	button?: React.ReactNode;
	/**
	 * Controls wether to show the error block or fail silently
	 */
	showErrorBlock?: boolean;
}

export interface BlockErrorProps extends BlockErrorBase {
	/**
	 * Error message to display below the content.
	 */
	errorMessage: React.ReactNode;
}

export type RenderErrorProps = {
	errorMessage: React.ReactNode;
};

export interface BlockErrorBoundaryProps extends BlockErrorBase {
	/**
	 * Override the default error with a function that takes the error message and returns a React component
	 */
	renderError?: ( props: RenderErrorProps ) => React.ReactNode;
	showErrorMessage?: boolean;
}

export interface DerivedStateReturn {
	errorMessage: JSX.Element | string;
	hasError: boolean;
}

export interface ReactError {
	status: string;
	statusText: string;
	message: string;
}
