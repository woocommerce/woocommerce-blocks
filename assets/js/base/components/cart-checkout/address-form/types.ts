/**
 * External dependencies
 */
import type {
	AddressField,
	AddressFields,
	AddressType,
	ShippingAddress,
	KeyedAddressField,
} from '@woocommerce/settings';

export type FieldConfig = Record<
	keyof AddressFields,
	Partial< AddressField >
>;

export type FieldType = keyof AddressFields;

export type AddressFormFields = {
	fields: KeyedAddressField[];
	type: AddressType;
	required: KeyedAddressField[];
	hidden: KeyedAddressField[];
};

export interface AddressFormProps {
	// Id for component.
	id?: string;
	// Unique id for form.
	instanceId: string;
	// Type of form (billing or shipping).
	type?: AddressType;
	// Array of fields in form.
	fields: FieldType[];
	// Field configuration for fields in form.
	fieldConfig?: FieldConfig;
	// Called with the new address data when the address form changes. This is only called when all required fields are filled and there are no validation errors.
	onChange: ( newValue: ShippingAddress ) => void;
	// Values for fields.
	values: ShippingAddress;
}
