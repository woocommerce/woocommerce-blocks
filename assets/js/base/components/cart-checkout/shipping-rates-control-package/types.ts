/**
 * External dependencies
 */
import type { ReactElement } from 'react';
import type { PackageRateOption } from '@woocommerce/type-defs/shipping';
import type { CartShippingPackageShippingRate } from '@woocommerce/type-defs/cart';

export interface PackageItem {
	name: string;
	key: string;
	quantity: number;
}

export interface Destination {
	address_1: string;
	address_2: string;
	city: string;
	state: string;
	postcode: string;
	country: string;
}

export interface PackageData {
	destination: Destination;
	name: string;
	shipping_rates: CartShippingPackageShippingRate[];
	items: PackageItem[];
}

export type PackageRateRenderOption = (
	option: CartShippingPackageShippingRate
) => PackageRateOption;

// A flag can be ternary if true, false, and undefined are all valid options.
// In our case, we use this for collapsible and showItems, having a boolean will force that
// option, having undefined will let the component decide the logic based on other factors.
export type TernaryFlag = boolean | undefined;

export interface PackageProps {
	/* PackageId can be a string, WooCommerce Subscriptions uses strings for example, but WooCommerce core uses numbers */
	packageId: string | number;
	renderOption?: PackageRateRenderOption | undefined;
	collapse?: boolean;
	packageData: PackageData;
	className?: string;
	collapsible?: TernaryFlag;
	noResultsMessage: ReactElement;
	showItems?: TernaryFlag;
}
