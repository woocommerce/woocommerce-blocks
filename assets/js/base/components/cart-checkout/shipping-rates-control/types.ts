/**
 * External dependencies
 */
import { CartResponseShippingRate } from '@woocommerce/type-defs/cart-response';
import type { ReactElement } from 'react';

/**
 * Internal dependencies
 */
import {
	PackageRateRenderOption,
	TernaryFlag,
} from '../shipping-rates-control-package';

export interface PackagesProps {
	// Array of packages
	packages: CartResponseShippingRate[];

	// If the package should be rendered as a collapsible panel
	collapsible?: TernaryFlag;

	// If we should items below the package name
	showItems?: TernaryFlag;

	// Rendered when there are no rates in a package
	noResultsMessage: ReactElement;

	// Function to render a shipping rate
	renderOption: PackageRateRenderOption;
}

export interface ShippingRatesControlProps {
	// If true, when multiple packages are rendered, you can see each package's items
	showItems?: TernaryFlag;

	// If true, when multiple packages are rendered they can be toggled open and closed
	collapsible?: TernaryFlag;

	// Array of packages containing shipping rates
	shippingRates: CartResponseShippingRate[];

	// Class name for package rates
	className?: string | undefined;

	// True when rates are being loaded
	isLoadingRates: boolean;

	// Rendered when there are no packages
	noResultsMessage: ReactElement;

	// Function to render a shipping rate
	renderOption?: PackageRateRenderOption | undefined;

	// String equal to the block name where the Slot is rendered
	context: 'woocommerce/cart' | 'woocommerce/checkout';
}
