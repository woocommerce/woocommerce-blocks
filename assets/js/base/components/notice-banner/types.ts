export interface NoticeBannerProps {
	// The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
	children: React.ReactNode;
	// Additional class name to give to the notice.
	className?: string | undefined;
	// Determines whether the notice can be dismissed by the user.
	isDismissible?: boolean | undefined;
	// Function called when dismissing the notice.
	onRemove?: ( () => void ) | undefined;
	// Determines the level of politeness for the notice when read by screen readers.
	politeness?: 'polite' | 'assertive' | undefined;
	// Optionally provided to change the spoken message for assistive technology.
	spokenMessage?: string | React.ReactNode | undefined;
	// Status determines the color of the notice and the icon.
	status: 'success' | 'error' | 'info' | 'warning' | 'default';
	// Summary text shown above passed notices/children.
	summary?: string | undefined;
}
