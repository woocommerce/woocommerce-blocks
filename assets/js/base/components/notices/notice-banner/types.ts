export interface NoticeBannerProps {
	className?: string | undefined;
	status: 'success' | 'error' | 'info' | 'warning' | 'default';
	politeness?: 'polite' | 'assertive' | undefined;
	isDismissible?: boolean | undefined;
	spokenMessage?: string | React.ReactNode | undefined;
	// Function called when dismissing the notice
	onRemove?: ( () => void ) | undefined;
	// The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
	children: React.ReactNode;
}
