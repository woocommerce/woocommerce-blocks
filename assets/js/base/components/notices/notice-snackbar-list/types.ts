export type NoticeProps = {
	id: string;
	content: string;
	status: 'success' | 'error' | 'info' | 'warning' | 'default';
	explicitDismiss?: boolean | undefined;
};

export type SnackbarListProps = {
	className?: string | undefined;
	notices: NoticeProps[];
	onRemove: ( id: string ) => void;
	children?: Array< React.ReactNode > | undefined;
};
