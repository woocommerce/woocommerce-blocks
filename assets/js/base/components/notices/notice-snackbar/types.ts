export interface NoticeSnackbarProps {
	className?: string | undefined;
	status: 'success' | 'error' | 'info' | 'warning' | 'default';
	politeness?: 'polite' | 'assertive' | undefined;
	explicitDismiss?: boolean | undefined;
	spokenMessage?: string | React.ReactNode | undefined;
	// Function called when dismissing the notice
	onRemove?: ( () => void ) | undefined;
	// The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
	children: React.ReactNode;
	/**
	 * A ref to the list that contains the snackbar.
	 */
	listRef?: React.MutableRefObject< HTMLDivElement | null > | undefined;
}
