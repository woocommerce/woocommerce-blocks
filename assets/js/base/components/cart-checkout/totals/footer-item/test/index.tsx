/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import { allSettings } from '@woocommerce/settings';
import { Currency } from '@woocommerce/types';
import { CartResponseTotals } from '@woocommerce/type-defs/cart-response';

/**
 * Internal dependencies
 */
import TotalsFooterItem from '../index';

describe( 'TotalsFooterItem', () => {
	beforeEach( () => {
		allSettings.taxesEnabled = true;
		allSettings.displayCartPricesIncludingTax = true;
	} );
	const currency: Currency = {
		code: 'GBP',
		decimalSeparator: '.',
		minorUnit: 2,
		prefix: '£',
		suffix: '',
		symbol: '£',
		thousandSeparator: ',',
	};

	const values: CartResponseTotals = {
		currency_code: 'GBP',
		currency_decimal_separator: '.',
		currency_minor_unit: 2,
		currency_prefix: '£',
		currency_suffix: '',
		currency_symbol: '£',
		currency_thousand_separator: ',',
		tax_lines: [],
		total_discount: '0',
		total_discount_tax: '0',
		total_fees: '0',
		total_fees_tax: '0',
		total_items: '7100',
		total_items_tax: '0',
		total_price: '8500',
		total_shipping: '0',
		total_shipping_tax: '0',
		total_tax: '0',
	};

	it( 'Does not show the "including %s of tax" line if tax is 0', () => {
		const { container } = render(
			<TotalsFooterItem currency={ currency } values={ values } />
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'Does not show the "including %s of tax" line if tax is disabled', () => {
		allSettings.taxesEnabled = false;
		/* This shouldn't ever happen if taxes are disabled, but this is to test whether the taxesEnabled setting works */
		const valuesWithTax = {
			...values,
			total_tax: '100',
			total_items_tax: '100',
		};
		const { container } = render(
			<TotalsFooterItem currency={ currency } values={ valuesWithTax } />
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'Shows the "including %s of tax" line if tax is greater than 0', () => {
		const valuesWithTax = {
			...values,
			total_tax: '100',
			total_items_tax: '100',
		};
		const { container } = render(
			<TotalsFooterItem currency={ currency } values={ valuesWithTax } />
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'Shows the "including %s TAX LABEL" line with single tax label', () => {
		const valuesWithTax = {
			...values,
			total_tax: '100',
			total_items_tax: '100',
			tax_lines: [ { name: '10% VAT', price: '100', rate: '10.000' } ],
		};
		const { container } = render(
			<TotalsFooterItem currency={ currency } values={ valuesWithTax } />
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'Shows the "including %s TAX LABELS" line with multiple tax labels', () => {
		const valuesWithTax = {
			...values,
			total_tax: '100',
			total_items_tax: '100',
			tax_lines: [
				{ name: '10% VAT', price: '50', rate: '10.000' },
				{ name: '5% VAT', price: '50', rate: '5.000' },
			],
		};
		const { container } = render(
			<TotalsFooterItem currency={ currency } values={ valuesWithTax } />
		);
		expect( container ).toMatchSnapshot();
	} );
} );
