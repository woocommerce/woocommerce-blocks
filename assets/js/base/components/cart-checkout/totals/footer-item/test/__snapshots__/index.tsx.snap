// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalsFooterItem Does not show the "including %s of tax" line if tax is 0 1`] = `
<div>
  <div
    class="wc-block-components-totals-item wc-block-components-totals-footer-item"
  >
    <span
      class="wc-block-components-totals-item__label"
    >
      Total
    </span>
    <span
      class="wc-block-formatted-money-amount wc-block-components-formatted-money-amount wc-block-components-totals-item__value"
    >
      £85.00
    </span>
    <div
      class="wc-block-components-totals-item__description"
    />
  </div>
</div>
`;

exports[`TotalsFooterItem Does not show the "including %s of tax" line if tax is disabled 1`] = `
<div>
  <div
    class="wc-block-components-totals-item wc-block-components-totals-footer-item"
  >
    <span
      class="wc-block-components-totals-item__label"
    >
      Total
    </span>
    <span
      class="wc-block-formatted-money-amount wc-block-components-formatted-money-amount wc-block-components-totals-item__value"
    >
      £85.00
    </span>
    <div
      class="wc-block-components-totals-item__description"
    />
  </div>
</div>
`;

exports[`TotalsFooterItem Shows the "including %s of tax" line if tax is greater than 0 1`] = `
<div>
  <div
    class="wc-block-components-totals-item wc-block-components-totals-footer-item"
  >
    <span
      class="wc-block-components-totals-item__label"
    >
      Total
    </span>
    <span
      class="wc-block-formatted-money-amount wc-block-components-formatted-money-amount wc-block-components-totals-item__value"
    >
      £85.00
    </span>
    <div
      class="wc-block-components-totals-item__description"
    >
      <p
        class="wc-block-components-totals-footer-item-tax"
      >
        Including 
        <span
          class="wc-block-formatted-money-amount wc-block-components-formatted-money-amount wc-block-components-totals-footer-item-tax-value"
        >
          £1.00
        </span>
         in taxes
      </p>
    </div>
  </div>
</div>
`;
